Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp

```diff
@@ -121,21 +122,24 @@ void WebPermissionController::tryProcessingRequests()
     }
 }
 
-void WebPermissionController::addObserver(WebCore::PermissionObserver& observer)
+void WebPermissionController::addObserver(WebCore::PermissionStatusIdentifier statusIdentifier, WebCore::ClientOrigin& origin, WebCore::PermissionDescriptor& descriptor, WebCore::ScriptExecutionContextIdentifier contextIdentifier)
 {
-    m_observers.add(observer);
+    ASSERT(isMainRunLoop());
+    m_observers.add(statusIdentifier, ObserverData { origin, descriptor, contextIdentifier });
 }
 
-void WebPermissionController::removeObserver(WebCore::PermissionObserver& observer)
+void WebPermissionController::removeObserver(WebCore::PermissionStatusIdentifier statusIdentifier)
 {
-    m_observers.remove(observer);
+    ASSERT(isMainRunLoop());
+    m_observers.remove(statusIdentifier);
 }
 
 void WebPermissionController::permissionChanged(const WebCore::ClientOrigin& origin, const WebCore::PermissionDescriptor& descriptor, WebCore::PermissionState state)
 {
+    ASSERT(isMainRunLoop());
     for (auto& observer : m_observers) {
```

Comment:
`for (auto& [permissionStatusIdentifier, observer] : m_observers) {`
would look better IMO.

Reply:
Fixed!